From c127a26ad3ad4fdf3c04a6de696696050b0818ec Mon Sep 17 00:00:00 2001
From: gollarx <ratnakumarix.golla@intel.com>
Date: Fri, 6 Jan 2023 11:43:22 +0530
Subject: [PATCH] Fix for Vts MultipleStartStopReset

The issue with MultipleStartStopReset is when we do
start()->reset()->start(), the state is in RUNNING state, so start() is
returning the C2_BAD_STATE.

Added the functionality for reset() method in mfx_c2_component for
fixing the issue with the VtsHalMediaC2 module.

Tracked-On: OAM-103591
Signed-off-by: gollarx <ratnakumarix.golla@intel.com>

diff --git a/c2_components/src/mfx_c2_component.cpp b/c2_components/src/mfx_c2_component.cpp
index 076dca1..0bb78f7 100755
--- a/c2_components/src/mfx_c2_component.cpp
+++ b/c2_components/src/mfx_c2_component.cpp
@@ -428,10 +428,23 @@ c2_status_t MfxC2Component::stop()
 c2_status_t MfxC2Component::reset()
 {
     MFX_DEBUG_TRACE_FUNC;
+    c2_status_t res = C2_OK;
+
     if(m_state == State::STOPPED)
-        return C2_OK;
-    else
-        return stop();
+        return res;
+    else {
+        res = DoStop(true);
+        {
+            std::lock_guard<std::mutex> lock(m_stateMutex);
+            if (C2_OK == res) { // DoStop succeeded
+                 m_state = m_nextState = State::STOPPED;
+             } else {
+                 m_nextState = m_state;
+             }
+             m_condStateStable.notify_all();
+         }
+         return res;
+     }
 }
 
 c2_status_t MfxC2Component::release()
-- 
2.34.1

